File: samples/UserDataExportChaserSample/UserDataExportChaser.cpp
//
// Copyright 2023 Autodesk
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

#include "ExportOptionsDialog.h"

#include <MaxUsd/Builders/JobContextRegistry.h>
#include <MaxUsd/Builders/SceneBuilderOptions.h>
#include <MaxUsd/Builders/USDSceneBuilderOptions.h>
#include <MaxUsd/Chaser/ExportChaser.h>
#include <MaxUsd/Chaser/ExportChaserRegistry.h>

#include <maxscript/mxsplugin/mxsCustomAttributes.h>

#include <boost/tokenizer.hpp>
#include <custattrib.h>
#include <icustattribcontainer.h>
#include <object.h>
#include <string.h>
using namespace std::string_literals;

// Data types supported by the UserDataExportChaserSample
enum class PropertyType : int
{
    USER_PROP = 1,  // "user"
    CUSTOM_DATA = 2 // "custom"
};

// The UserDataExportChaserSample class
// This Export Chaser sample adds Custom Data to the exported prims if the specified user
// property or custom attribute are found in the corresponding Node. The sample chaser
// acts with default data or using the arguments that may have been passed to the chaser.
class UserDataExportChaserSample : public pxr::MaxUsdExportChaser
{
public:
    // Constructor - can be customized for the Export Chaser requirements.
    // In the provided sample it receives the stage and Prims to Node map, which are
    // the minimal arguments for a functional chaser, and the chaser arguments, which are
    // not mandatory but can be used to parameterize the chaser.
    UserDataExportChaserSample(
        pxr::UsdStagePtr                                                      stage,
        const pxr::MaxUsdExportChaserRegistry::FactoryContext::PrimToNodeMap& primToNodeMap,
        const MaxUsd::USDSceneBuilderOptions::ChaserArgs&                     args,
        const pxr::VtDictionary&                                              jobContextOptions);

    // Processing that needs to run after the main 3ds Max USD export loop.
    bool PostExport() override;

private:
    // The exported stage
    pxr::UsdStagePtr stage;
    // Maps full USD prim paths to INodes
    const pxr::MaxUsdExportChaserRegistry::FactoryContext::PrimToNodeMap& primToNodeMap;
    // potentially user-customized jobContext options.
    const pxr::VtDictionary& jobContextOptions;

    // sample specific parameters
    // maps the data to be exported as Custom Data for the prims
    typedef std::map<PropertyType, std::set<std::string>> PropertyMap;
    PropertyMap                                           dataToExport;
};

// Add the using statement here to make the export chaser factory macro work
PXR_NAMESPACE_USING_DIRECTIVE
// Macro to register the Export Chaser. Defines a factory method for the chaser name. The 'ctx' will
// be of type MaxUsdExportChaserRegistry::FactoryContext. The method should return a
// MaxUsdExportChaser*. There are no guarantees about the lifetime of 'ctx'. It is also very
// important to set the project option "Remove unreferenced code and data" to NO, not doing so could
// cause the Macro to be optimized out and the Export Chaser to never be properly registered.
PXR_MAXUSD_DEFINE_EXPORT_CHASER_FACTORY(
    UserData,
    "User Data C++ DEMO",
    "Chaser to export user data along the exported USD prims",
    ctx)
{
    // Fetching the export chaser parameters
    // The chaser arguments are mapped using their registration chaser name (first arg of the macro)
    MaxUsd::USDSceneBuilderOptions::ChaserArgs myArgs;
    TfMapLookup(ctx.GetJobArgs().GetAllChaserArgs(), "UserData", &myArgs);
    const pxr::VtDictionary& jobContextOptions
        = ctx.GetJobArgs().GetJobContextOptions(TfToken("CustomDemoContext"));
    return new UserDataExportChaserSample(
        ctx.GetStage(), ctx.GetPrimToNodeMap(), myArgs, jobContextOptions);
}

UserDataExportChaserSample::UserDataExportChaserSample(
    pxr::UsdStagePtr                                                      stage,
    const pxr::MaxUsdExportChaserRegistry::FactoryContext::PrimToNodeMap& primToNodeMap,
    const MaxUsd::USDSceneBuilderOptions::ChaserArgs&                     args,
    const pxr::VtDictionary&                                              jobContextOptions)
    : stage(stage)
    , primToNodeMap(primToNodeMap)
    , jobContextOptions(jobContextOptions)
{
    // Default configuration
    dataToExport
        = { { PropertyType::USER_PROP, { "myUserProperty" } }, { PropertyType::CUSTOM_DATA, {} } };

    // Parsing the specific export chaser arguments
    boost::char_separator<char> sep(",");
    for (std::pair<std::string, std::string> item : args) {
        if (item.first == "user") {
            dataToExport[PropertyType::USER_PROP].clear();
            boost::tokenizer<boost::char_separator<char>> tokens(item.second, sep);
            for (auto name : tokens) {
                dataToExport[PropertyType::USER_PROP].insert(name);
            }
        } else if (item.first == "custom") {
            dataToExport[PropertyType::CUSTOM_DATA].clear();
            boost::tokenizer<boost::char_separator<char>> tokens(item.second, sep);
            for (auto name : tokens) {
                dataToExport[PropertyType::CUSTOM_DATA].insert(name);
            }
        } else {
            TF_WARN(
                "Wrong user data type ('%s') passed as argument to UserPropertyExportChaser",
                item.first.c_str());
        }
    }

    if (!jobContextOptions.empty()) {
        if (auto greetUser = jobContextOptions.GetValueAtPath({ "Greeting"s, "Greet User"s })) {
            if (greetUser->Get<bool>()) {
                std::string username;
                if (auto userNameV
                    = jobContextOptions.GetValueAtPath({ "Greeting"s, "User Name"s })) {
                    if (userNameV->CanCast<std::string>()) {
                        username = userNameV->Get<std::string>();
                    }
                }
                bool formal = false;
                if (auto formalV = jobContextOptions.GetValueAtPath({ "Greeting"s, "Formal"s })) {
                    if (formalV->CanCast<bool>()) {
                        formal = formalV->Get<bool>();
                    }
                }
                std::string option;
                if (auto optionAV = jobContextOptions.GetValueAtPath({ "Option A"s })) {
                    if (optionAV->CanCast<bool>() && optionAV->Get<bool>()) {
                        option = "Option A";
                    }
                }
                bool optionB = false;
                if (auto optionBV = jobContextOptions.GetValueAtPath({ "Option B"s })) {
                    if (optionBV->CanCast<bool>() && optionBV->Get<bool>()) {
                        option = "Option B";
                    }
                }

                std::string greeting;
                if (formal) {
                    time_t now = time(0);
                    tm*    localtm = localtime(&now);
                    greeting = (localtm->tm_hour > 12) ? "Good Afternoon" : "Good Morning";
                } else {
                    greeting = "Hi";
                }
                static constexpr const char* message = "{0} '{1}' - You have chosen {2}";
                MaxUsd::Log::Info(message, greeting.c_str(), username.c_str(), option.c_str());
            }
        }
    }
}

bool UserDataExportChaserSample::PostExport()
{
    // Cycle through the exported prim/node
    for (std::pair<SdfPath, INode*> exported : primToNodeMap) {
        pxr::UsdPrim prim = stage->GetPrimAtPath(exported.first);
        INode*       node = exported.second;

        // Are any of the specified user properties found in the Node
        for (std::string propName : dataToExport[PropertyType::USER_PROP]) {
            auto mxsPropName = MaxUsd::UsdStringToMaxString(propName);
            if (node->UserPropExists(mxsPropName)) {
                TSTR value;
                node->GetUserPropString(mxsPropName, value);

                if (_tcsicmp(value, _T("true")) == 0) {
                    prim.SetCustomDataByKey(pxr::TfToken(propName), pxr::VtValue(true));
                } else if (_tcsicmp(value, _T("false")) == 0) {
                    prim.SetCustomDataByKey(pxr::TfToken(propName), pxr::VtValue(false));
                } else {
                    // Note - to simplify the chaser sample implementation, the user property
                    // conversion to numerical values was not completed. It will be written as a
                    // simple string. The same chaser written in Python properly handles this
                    // situation however.
                    auto valueString = MaxUsd::MaxStringToUsdString(value);
                    // Strip the quoting marks from the string (if the property was originally a
                    // string)
                    if (*valueString.begin() == '\"' && *(valueString.end() - 1) == '\"') {
                        valueString.erase(valueString.begin());
                        valueString.erase(valueString.end() - 1);
                    }
                    prim.SetCustomDataByKey(pxr::TfToken(propName), pxr::VtValue(valueString));
                }
            }
        }

        // For the purpose of the sample, the chaser finds the custom attributes only on
        // the base level, not on the modifiers or materials of the node
        BaseObject*           obj = node->GetObjectRef();
        ICustAttribContainer* cc = obj->GetCustAttribContainer();
        if (!cc) {
            continue;
        }
        // Are any of the specified custom attributes found in the Node
        for (std::string propName : dataToExport[PropertyType::CUSTOM_DATA]) {
            auto mxsPropName = MaxUsd::UsdStringToMaxString(propName);
            for (int i = 0; i < cc->GetNumCustAttribs(); i++) {
                CustAttrib* ca = cc->GetCustAttrib(i);
                if (!ca) {
                    continue;
                }
                MSCustAttrib* msCA = (MSCustAttrib*)ca->GetInterface(I_SCRIPTEDCUSTATTRIB);
                if (!msCA) {
                    continue;
                }

                for (int j = 0; j < ca->NumParamBlocks(); ++j) {
                    IParamBlock2* pb2 = ca->GetParamBlock(j);
                    if (!pb2) {
                        continue;
                    }
                    ParamID paramId = MaxUsd::FindParamId(pb2, mxsPropName);
                    // If this specific attribute doesn't exist, we skip to look for it in the next
                    // PB
                    if (paramId == -1) {
                        continue;
                    }

                    // Found the attribute, extract the value and write the data
                    {
                        const ParamDef& paramDef = pb2->GetParamDef(paramId);
                        Interval        iv = FOREVER;
                        bool            hasFoundParam = false;
                        switch (paramDef.type) {
                        case TYPE_BOOL: {
                            int boolVal = 0;
                            hasFoundParam = pb2->GetValue(paramId, 0, boolVal, iv);
                            prim.SetCustomDataByKey(
                                pxr::TfToken(propName),
                                pxr::VtValue((boolVal == 0) ? false : true));
                            break;
                        }
                        case TYPE_INT: {
                            int strVal = false;
                            hasFoundParam = pb2->GetValue(paramId, 0, strVal, iv);
                            prim.SetCustomDataByKey(pxr::TfToken(propName), pxr::VtValue(strVal));
                            break;
                        }
                        case TYPE_FLOAT: {
                            float strVal = false;
                            hasFoundParam = pb2->GetValue(paramId, 0, strVal, iv);
                            prim.SetCustomDataByKey(pxr::TfToken(propName), pxr::VtValue(strVal));
                            break;
                        }
                        case TYPE_STRING: {
                            const wchar_t* strVal = L"";
                            hasFoundParam = pb2->GetValue(paramId, 0, strVal, iv);
                            prim.SetCustomDataByKey(
                                pxr::TfToken(propName),
                                pxr::VtValue(MaxUsd::MaxStringToUsdString(strVal)));
                            break;
                        }
                        default:
                            TF_WARN(
                                "Unsupported Custom Attribute type for '%s' in "
                                "UserPropertyExportChaser",
                                propName.c_str());
                            break;
                        }
                        // No need to search further, the attribute was found
                        break;
                    }
                }
            }
        }
    }
    return true;
}

REGISTER_EXPORT_JOB_CONTEXT_FCT(
    CustomDemoContext,
    "Custom Context C++ DEMO",
    "Custom plug-in configuration")
{
    VtDictionary extraArgs;
    extraArgs[MaxUsdSceneBuilderOptionsTokens->chaserNames]
        = VtValue(std::vector<VtValue> { VtValue(std::string("UserData")) });
    VtValue chaserArgUserProp(
        std::vector<VtValue> { VtValue(std::string("UserData")),
                               VtValue(std::string("user")),
                               VtValue(std::string("myUserFloatProperty,myUserProperty")) });
    VtValue chaserArgCustomProp(std::vector<VtValue> { VtValue(std::string("UserData")),
                                                       VtValue(std::string("custom")),
                                                       VtValue(std::string("inGame")) });
    extraArgs[MaxUsdSceneBuilderOptionsTokens->chaserArgs]
        = VtValue(std::vector<VtValue> { chaserArgUserProp, chaserArgCustomProp });

    return extraArgs;
}

TF_REGISTRY_FUNCTION(MaxUsdJobContextRegistry)
{
    MaxUsdJobContextRegistry::GetInstance().SetExportOptionsUI(
        "CustomDemoContext", &ExportOptionsDialog::ShowOptionsDialog);
}

File: samples/UserDataExportChaserSample/Contents/python/UserDataExportChaser.py
#
# Copyright 2023 Autodesk
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
import maxUsd

from pxr import Usd
from pymxs import runtime as mxs

import os, sys
from enum import Enum
import traceback

# Export Chaser sample that adds Custom Data to the exported prims if the specified user
# property or custom attribute are found in the corresponding Node. The sampled chaser
# acts with default data or using the arguments that may have been passed to the chaser.

# the supported data types by the UserDataExportChaserSample
class PropertyType(Enum):
    USER_PROP = 1
    CUSTOM_DATA = 2


class userDataExportChaserSample(maxUsd.ExportChaser):
    
    default_property = { PropertyType.USER_PROP : ["myUserProperty"], PropertyType.CUSTOM_DATA : [] }

	# constructor - can be customized for the Export Chaser requirements
	# in the provided sample, it receives the stage and Prims to Node map, which are 
	# the minimal arguments for a functional chaser, and the chaser arguments, which are
	# not mandatory but can be use to parametrize the chaser.
    def __init__(self, factoryContext, *args, **kwargs):
        super(userDataExportChaserSample, self).__init__(factoryContext, *args, **kwargs)

        self.properties = userDataExportChaserSample.default_property

        # retrieve the specified arguments to be used with the export chaser
        # the chaser's arguments can be specified for different chasers
        # the returned arguments are in the form of a dictionary
        if 'UserData' in factoryContext.GetJobArgs().GetAllChaserArgs():
            chaser_args = factoryContext.GetJobArgs().GetAllChaserArgs()['UserData']
            for type, names in chaser_args.items():
                if type =="user":
                    self.properties[PropertyType.USER_PROP] = []
                    for name in names.split(','):
                        self.properties[PropertyType.USER_PROP].append(name)
                elif type == "custom":
                    self.properties[PropertyType.CUSTOM_DATA] = []
                    for name in names.split(','):
                        self.properties[PropertyType.CUSTOM_DATA].append(name)

        # retrieve the export mapping dictionary for the stage
        # the dictionary maps the resulting USD prim paths to their MAXScript NodeHandles origin
        self.primsToNodeHandles = factoryContext.GetPrimsToNodeHandles()

        # retrieve the USD stage being written to
        self.stage = factoryContext.GetStage()

        # retrieve the jobContextOptions the user may have modified through the UI
        self.contextOptions = factoryContext.GetJobArgs().GetJobContextOptions("CustomDemoContext")

	# processing that needs to run after the main 3ds Max USD export loop.
    def PostExport(self):
        
        # the jobContextOptions the user may have modified through the UI 
        if self.contextOptions.get('do_something_special', False):
            # do something Special
            pass

        try:
            for prim_path, node_handle in self.primsToNodeHandles.items():
                node = mxs.maxOps.getNodeByHandle(node_handle)
                prim = self.stage.GetPrimAtPath(prim_path)
                
                for type, names in self.properties.items():
                    for name in names:
                        if type == PropertyType.USER_PROP:
                            prop = mxs.getUserProp(node, name)
                            # verify the property exists
                            if prop == None:
                                continue
                            if mxs.classOf(prop) == mxs.String:
                                # strip the quoting marks from the string
                                prop = prop[1:len(prop)-1]
                            prim.SetCustomDataByKey(name, prop)
                
                        elif type == PropertyType.CUSTOM_DATA:
                            if mxs.isProperty(node, name):
                                prop = mxs.getProperty(node, name)
                                prim.SetCustomDataByKey(name, prop)
                
        except Exception as e:
            # Quite useful to debug errors in a Python callback
            print('Write() - Error: %s' % str(e))
            print(traceback.format_exc())
            return False
        
        return True


# Job Contexts allow to customize several options during a maxUsd export
#   - 'chaserNames'
#   - 'chaserArgs'
#   - 'convertMaterialsTo'
#
# This sample will add a "Custom Context Demo" option in the "PlugIn configuration" drop down in maxUsd exports
def CustomExportJobContextSample():
    # build a dictionary of the options to set using the context
    extraArgs = {}
    
    # The UserData chaser needs to be enabled in this job context
    extraArgs['chaserNames']  = ['UserData']
    extraArgs['chaserArgs'] = [['UserData', 'user', 'myUserFloatProperty,myUserProperty'], ['UserData', 'custom', 'inGame']]
    return extraArgs



maxUsd.ExportChaser.Register(userDataExportChaserSample, "UserData", "User Data Python DEMO", "Chaser to export user data along the exported USD prims")

maxUsd.JobContextRegistry.RegisterExportJobContext("CustomDemoContext", "Custom Context Python DEMO", "Custom plug-in configuration", CustomExportJobContextSample)

# Uncomment code below to export a scene using the export chaser sample
# The code prepares the exporter options to
#  - export to a USD file in ASCII
#  - sets the ChaserNames to only use 'UserData'; the name of export chaser sample
#  - sets the arguments for the 'UserData' export chaser
#    - the list of user property ('user') to export as USD Custom Data; comma separated property names
#    - the list of custom attribute ('custom') to export as USD Custom Data; comma separated property names

#opt = mxs.USDExporter.CreateOptions()
#opt.FileFormat = mxs.Name("ascii")
#opt.RootPrimPath = "/"
#opt.ChaserNames = mxs.Array("UserData")
#opt.AllChaserArgs = mxs.Array("UserData", "user", "myUserFloatProperty,myUserProperty", "UserData", "custom", "inGame")
#mxs.USDExporter.ExportFile( "c:\\temp\\myscene_with_custom_data.usd",
#                            exportOptions=opt)

def showJobExportOptions(jobName, parentUI, contextOptions):
    # 3dsMax 2025 and later is using PySide6, while 3dsMax 2024 and earlier is
    # using PySide2 - so we need to check for both
    try: 
        import PySide6.QtWidgets as QtWidgets
    except ImportError:
        try: 
            import PySide2.QtWidgets as QtWidgets
        except ImportError as err:
            print(f"Could not import QtWidgets from neither PySide6 nor PySide2 {err=}.")
            raise err
    
    try: 
        dlg = QtWidgets.QDialog(parentUI)
        dlg.setWindowTitle(f"Export Options for {jobName}")
        dlg.setModal(True)
        dlg.setLayout(QtWidgets.QVBoxLayout())

        checkBoxSpecial = QtWidgets.QCheckBox("Do something Special")
        checkBoxSpecial.setChecked(contextOptions.get('do_something_special', False))
        dlg.layout().addWidget(checkBoxSpecial)

        # ...

        buttonBox = QtWidgets.QDialogButtonBox(QtWidgets.QDialogButtonBox.Ok | QtWidgets.QDialogButtonBox.Cancel)
        buttonBox.accepted.connect(dlg.accept)
        buttonBox.rejected.connect(dlg.reject)
        dlg.layout().addWidget(buttonBox, 0)

        if dlg.exec() == QtWidgets.QDialog.Accepted:
            contextOptions['do_something_special'] = checkBoxSpecial.isChecked()
            #...

    except Exception as err:
        print(f"Unexpected {err=}, {type(err)=}")

    return contextOptions

maxUsd.JobContextRegistry.SetExportOptionsUI("CustomDemoContext", showJobExportOptions)
